File: src/maden_lexer.py
from typing import List


class TokenType:
    FUNGSI = "FUNGSI"
    KEMBALI = "KEMBALI"
    CETAK = "CETAK"
    JIKA = "JIKA"
    SELAIN = "SELAIN"
    ULANGI = "ULANGI"
    UNTUK = "UNTUK"
    SETIAP = "SETIAP"
    DALAM = "DALAM"
    SELAMA = "SELAMA"
    IMPOR = "IMPOR"
    BENAR = "BENAR"
    SALAH = "SALAH"
    KOSONG = "KOSONG"
    DAN = "DAN"
    ATAU = "ATAU"
    NOT = "NOT"

    INT8 = "INT8"
    INT128 = "INT128"
    ACCELERATOR = "ACCELERATOR"
    PRECISION = "PRECISION"
    PARALLEL = "PARALLEL"

    IDENTIFIER = "IDENTIFIER"
    ANGKA = "ANGKA"
    TEKS = "TEKS"
    KURUNG_BUKA = "KURUNG_BUKA"
    KURUNG_TUTUP = "KURUNG_TUTUP"
    KURUNG_KOTAK_BUKA = "KURUNG_KOTAK_BUKA"
    KURUNG_KOTAK_TUTUP = "KURUNG_KOTAK_TUTUP"
    KURUNG_KURUNG_BUKA = "KURUNG_KURUNG_BUKA"
    KURUNG_KURUNG_TUTUP = "KURUNG_KURUNG_TUTUP"
    KOMA = "KOMA"
    TITIK_DUA = "TITIK_DUA"
    TITIK_KOMA = "TITIK_KOMA"
    SAMA_DENGAN = "SAMA_DENGAN"
    TAMBAH = "TAMBAH"
    KURANG = "KURANG"
    KALI = "KALI"
    BAGI = "BAGI"
    MODULUS = "MODULUS"
    LEBIH_BESAR = "LEBIH_BESAR"
    LEBIH_KECIL = "LEBIH_KECIL"
    LEBIH_BESAR_SAMA = "LEBIH_BESAR_SAMA"
    LEBIH_KECIL_SAMA = "LEBIH_KECIL_SAMA"
    SAMA_DENGAN_DUA = "SAMA_DENGAN_DUA"
    TIDAK_SAMA = "TIDAK_SAMA"
    KOMENTAR = "KOMENTAR"
    NEWLINE = "NEWLINE"
    EOF = "EOF"


class Token:
    def __init__(self, tipe: str, nilai, baris: int, kolom: int):
        self.tipe = tipe
        self.nilai = nilai
        self.baris = baris
        self.kolom = kolom

    def __repr__(self):
        return f"Token({self.tipe}, {self.nilai!r}, baris={self.baris})"


class Lexer:
    def __init__(self, kode: str):
        self.kode = kode
        self.pos = 0
        self.baris = 1
        self.kolom = 1
        self.token = []
        self.keyword = {
            "fungsi": TokenType.FUNGSI, "kembali": TokenType.KEMBALI,
            "cetak": TokenType.CETAK, "jika": TokenType.JIKA,
            "selain": TokenType.SELAIN, "ulangi": TokenType.ULANGI,
            "untuk": TokenType.UNTUK, "setiap": TokenType.SETIAP,
            "dalam": TokenType.DALAM, "selama": TokenType.SELAMA,
            "import": TokenType.IMPOR, "benar": TokenType.BENAR,
            "salah": TokenType.SALAH, "kosong": TokenType.KOSONG,
            "dan": TokenType.DAN, "atau": TokenType.ATAU,
            "not": TokenType.NOT, "tidak": TokenType.NOT,
        }
        self.annotations = {
            "@int8": TokenType.INT8, "@int128": TokenType.INT128,
            "@accelerator": TokenType.ACCELERATOR,
            "@precision": TokenType.PRECISION, "@parallel": TokenType.PARALLEL,
        }

    def tokenisasi(self) -> List[Token]:
        while self.pos < len(self.kode):
            char = self.kode[self.pos]
            if char in " \t\r":
                self.pos += 1
                self.kolom += 1
                continue
            if char == "\n":
                self.token.append(Token(TokenType.NEWLINE, "\n", self.baris, self.kolom))
                self.pos += 1
                self.baris += 1
                self.kolom = 1
                continue
            if char == "/" and self._ikuti("/"):
                self.pos += 2
                self.kolom += 2
                while self.pos < len(self.kode) and self.kode[self.pos] != "\n":
                    self.pos += 1
                    self.kolom += 1
                continue
            if char in "\"'":
                self._token_teks(char)
                continue
            # Tanda minus selalu menjadi token operator. Unary minus ditangani parser.
            if char.isdigit():
                self._token_angka()
                continue
            if char.isalpha() or char == "_":
                self._token_identifier()
                continue
            if char == "@":
                self._token_anotasi()
                continue

            operator_dua = {
                "==": (TokenType.SAMA_DENGAN_DUA, "=="),
                "!=": (TokenType.TIDAK_SAMA, "!="),
                ">=": (TokenType.LEBIH_BESAR_SAMA, ">="),
                "<=": (TokenType.LEBIH_KECIL_SAMA, "<="),
                "&&": (TokenType.DAN, "&&"),
                "||": (TokenType.ATAU, "||"),
            }
            pasangan = self.kode[self.pos:self.pos + 2]
            if pasangan in operator_dua:
                tipe, nilai = operator_dua[pasangan]
                self.token.append(Token(tipe, nilai, self.baris, self.kolom))
                self.pos += 2
                self.kolom += 2
                continue

            simbol = {
                "(": TokenType.KURUNG_BUKA, ")": TokenType.KURUNG_TUTUP,
                "[": TokenType.KURUNG_KOTAK_BUKA, "]": TokenType.KURUNG_KOTAK_TUTUP,
                "{": TokenType.KURUNG_KURUNG_BUKA, "}": TokenType.KURUNG_KURUNG_TUTUP,
                ",": TokenType.KOMA, ":": TokenType.TITIK_DUA,
                ";": TokenType.TITIK_KOMA, "=": TokenType.SAMA_DENGAN,
                "+": TokenType.TAMBAH, "-": TokenType.KURANG,
                "*": TokenType.KALI, "/": TokenType.BAGI,
                "%": TokenType.MODULUS, ">": TokenType.LEBIH_BESAR,
                "<": TokenType.LEBIH_KECIL,
            }
            if char in simbol:
                self.token.append(Token(simbol[char], char, self.baris, self.kolom))
                self.pos += 1
                self.kolom += 1
                continue
            raise SyntaxError(f"Karakter tidak dikenal: '{char}' di baris {self.baris}, kolom {self.kolom}")

        self.token.append(Token(TokenType.EOF, None, self.baris, self.kolom))
        return self.token

    def _ikuti(self, nilai: str) -> bool:
        return self.pos + 1 < len(self.kode) and self.kode[self.pos + 1] == nilai

    def _token_teks(self, tanda: str):
        mulai_baris, mulai_kolom = self.baris, self.kolom
        self.pos += 1
        self.kolom += 1
        teks = []
        while self.pos < len(self.kode) and self.kode[self.pos] != tanda:
            if self.kode[self.pos] == "\n":
                self.baris += 1
                self.kolom = 1
            teks.append(self.kode[self.pos])
            self.pos += 1
            self.kolom += 1
        if self.pos >= len(self.kode):
            raise SyntaxError(f"Teks tidak ditutup di baris {mulai_baris}")
        self.pos += 1
        self.kolom += 1
        self.token.append(Token(TokenType.TEKS, "".join(teks), mulai_baris, mulai_kolom))

    def _token_angka(self):
        mulai_baris, mulai_kolom = self.baris, self.kolom
        awal = self.pos
        while self.pos < len(self.kode) and self.kode[self.pos].isdigit():
            self.pos += 1
            self.kolom += 1
        if self.pos < len(self.kode) and self.kode[self.pos] == ".":
            self.pos += 1
            self.kolom += 1
            while self.pos < len(self.kode) and self.kode[self.pos].isdigit():
                self.pos += 1
                self.kolom += 1
        teks = self.kode[awal:self.pos]
        nilai = float(teks) if "." in teks else int(teks)
        self.token.append(Token(TokenType.ANGKA, nilai, mulai_baris, mulai_kolom))

    def _token_identifier(self):
        mulai_baris, mulai_kolom = self.baris, self.kolom
        awal = self.pos
        while self.pos < len(self.kode) and (self.kode[self.pos].isalnum() or self.kode[self.pos] == "_"):
            self.pos += 1
            self.kolom += 1
        asli = self.kode[awal:self.pos]
        identitas = asli.lower()
        tipe = self.keyword.get(identitas, TokenType.IDENTIFIER)
        self.token.append(Token(tipe, identitas if tipe != TokenType.IDENTIFIER else asli, mulai_baris, mulai_kolom))

    def _token_anotasi(self):
        mulai_baris, mulai_kolom = self.baris, self.kolom
        awal = self.pos
        self.pos += 1
        self.kolom += 1
        while self.pos < len(self.kode) and (self.kode[self.pos].isalnum() or self.kode[self.pos] == "_"):
            self.pos += 1
            self.kolom += 1
        anotasi = self.kode[awal:self.pos].lower()
        tipe = self.annotations.get(anotasi)
        if tipe is None:
            raise SyntaxError(f"Anotasi tidak dikenal: '{anotasi}' di baris {mulai_baris}")
        self.token.append(Token(tipe, anotasi, mulai_baris, mulai_kolom))

File: src-js/maden-lexer.js
const TokenType = {
  FUNGSI: "FUNGSI",
  KEMBALI: "KEMBALI",
  CETAK: "CETAK",
  JIKA: "JIKA",
  SELAIN: "SELAIN",
  ULANGI: "ULANGI",
  SELAMA: "SELAMA",
  UNTUK: "UNTUK",
  SETIAP: "SETIAP",
  DALAM: "DALAM",
  BENAR: "BENAR",
  SALAH: "SALAH",
  KOSONG: "KOSONG",
  DAN: "DAN",
  ATAU: "ATAU",
  TIDAK: "TIDAK",

  IDENTIFIER: "IDENTIFIER",
  ANGKA: "ANGKA",
  TEKS: "TEKS",

  TAMBAH: "TAMBAH",
  KURANG: "KURANG",
  KALI: "KALI",
  BAGI: "BAGI",
  MOD: "MOD",

  KURANG_BESAR: "KURANG_BESAR",
  KURANG_KECIL: "KURANG_KECIL",
  LEBIH_BESAR: "LEBIH_BESAR",
  LEBIH_KECIL: "LEBIH_KECIL",
  LEBIH_BESAR_SAMA: "LEBIH_BESAR_SAMA",
  LEBIH_KECIL_SAMA: "LEBIH_KECIL_SAMA",
  SAMA_DENGAN: "SAMA_DENGAN",
  TIDAK_SAMA: "TIDAK_SAMA",

  KURUNG_BUKA: "KURUNG_BUKA",
  KURUNG_TUTUP: "KURUNG_TUTUP",
  BLOK_BUKA: "BLOK_BUKA",
  BLOK_TUTUP: "BLOK_TUTUP",
  KOMA: "KOMA",
  TITIK_KOMA: "TITIK_KOMA",
  SAMA: "SAMA",

  NEWLINE: "NEWLINE",
  EOF: "EOF",
};

class Token {
  constructor(type, value, line, column) {
    this.type = type;
    this.value = value;
    this.line = line;
    this.column = column;
  }
}

function tokenize(source) {
  const tokens = [];
  let i = 0;
  let line = 1;
  let column = 1;

  const keywords = {
    fungsi: TokenType.FUNGSI,
    kembali: TokenType.KEMBALI,
    cetak: TokenType.CETAK,
    jika: TokenType.JIKA,
    selain: TokenType.SELAIN,
    ulangi: TokenType.ULANGI,
    selama: TokenType.SELAMA,
    untuk: TokenType.UNTUK,
    setiap: TokenType.SETIAP,
    dalam: TokenType.DALAM,
    benar: TokenType.BENAR,
    salah: TokenType.SALAH,
    kosong: TokenType.KOSONG,
    dan: TokenType.DAN,
    atau: TokenType.ATAU,
    tidak: TokenType.TIDAK,
    not: TokenType.TIDAK,
  };

  while (i < source.length) {
    const ch = source[i];

    if (ch === ' ' || ch === '\t' || ch === '\r') {
      i++; column++; continue;
    }

    if (ch === '\n') {
      tokens.push(new Token(TokenType.NEWLINE, '\n', line, column));
      i++; line++; column = 1; continue;
    }

    if (ch === '"' || ch === "'") {
      const quote = ch;
      let value = '';
      i++; column++;
      while (i < source.length && source[i] !== quote) {
        if (source[i] === '\\') {
          i++; column++;
          if (i >= source.length) {
            throw new Error(`String tidak ditutup di baris ${line}, kolom ${column}`);
          }
          const escaped = source[i];
          value += {
            n: '\n',
            t: '\t',
            r: '\r',
            '\\': '\\',
            '"': '"',
            "'": "'",
          }[escaped] ?? escaped;
          i++; column++; continue;
        }
        value += source[i];
        i++; column++;
      }
      if (i >= source.length) {
        throw new Error(`String tidak ditutup di baris ${line}, kolom ${column}`);
      }
      i++; column++;
      tokens.push(new Token(TokenType.TEKS, value, line, column));
      continue;
    }

    if (/\d/.test(ch)) {
      let value = '';
      while (i < source.length && /\d/.test(source[i])) {
        value += source[i];
        i++; column++;
      }
      if (source[i] === '.' && /\d/.test(source[i + 1])) {
        value += '.';
        i++; column++;
        while (i < source.length && /\d/.test(source[i])) {
          value += source[i];
          i++; column++;
        }
      }
      tokens.push(new Token(TokenType.ANGKA, Number(value), line, column));
      continue;
    }

    if (/[A-Za-z_]/.test(ch)) {
      let value = '';
      while (i < source.length && /[A-Za-z0-9_]/.test(source[i])) {
        value += source[i];
        i++; column++;
      }
      const type = keywords[value] ?? TokenType.IDENTIFIER;
      tokens.push(new Token(type, value, line, column));
      continue;
    }

    if (source.startsWith('>=', i)) {
      tokens.push(new Token(TokenType.LEBIH_BESAR_SAMA, '>=', line, column));
      i += 2; column += 2; continue;
    }
    if (source.startsWith('<=', i)) {
      tokens.push(new Token(TokenType.LEBIH_KECIL_SAMA, '<=', line, column));
      i += 2; column += 2; continue;
    }
    if (source.startsWith('==', i)) {
      tokens.push(new Token(TokenType.SAMA_DENGAN, '==', line, column));
      i += 2; column += 2; continue;
    }
    if (source.startsWith('!=', i)) {
      tokens.push(new Token(TokenType.TIDAK_SAMA, '!=', line, column));
      i += 2; column += 2; continue;
    }
    if (source.startsWith('&&', i)) {
      tokens.push(new Token(TokenType.DAN, '&&', line, column));
      i += 2; column += 2; continue;
    }
    if (source.startsWith('||', i)) {
      tokens.push(new Token(TokenType.ATAU, '||', line, column));
      i += 2; column += 2; continue;
    }

    if (ch === '(') {
      tokens.push(new Token(TokenType.KURUNG_BUKA, ch, line, column));
      i++; column++; continue;
    }
    if (ch === ')') {
      tokens.push(new Token(TokenType.KURUNG_TUTUP, ch, line, column));
      i++; column++; continue;
    }
    if (ch === '{') {
      tokens.push(new Token(TokenType.BLOK_BUKA, ch, line, column));
      i++; column++; continue;
    }
    if (ch === '}') {
      tokens.push(new Token(TokenType.BLOK_TUTUP, ch, line, column));
      i++; column++; continue;
    }
    if (ch === '[') {
      tokens.push(new Token(TokenType.KURUNG_BUKA, ch, line, column));
      i++; column++; continue;
    }
    if (ch === ']') {
      tokens.push(new Token(TokenType.KURUNG_TUTUP, ch, line, column));
      i++; column++; continue;
    }
    if (ch === ',') {
      tokens.push(new Token(TokenType.KOMA, ch, line, column));
      i++; column++; continue;
    }
    if (ch === ';') {
      tokens.push(new Token(TokenType.TITIK_KOMA, ch, line, column));
      i++; column++; continue;
    }
    if (ch === '+') {
      tokens.push(new Token(TokenType.TAMBAH, ch, line, column));
      i++; column++; continue;
    }
    if (ch === '-') {
      tokens.push(new Token(TokenType.KURANG, ch, line, column));
      i++; column++; continue;
    }
    if (ch === '*') {
      tokens.push(new Token(TokenType.KALI, ch, line, column));
      i++; column++; continue;
    }
    if (ch === '/') {
      tokens.push(new Token(TokenType.BAGI, ch, line, column));
      i++; column++; continue;
    }
    if (ch === '%') {
      tokens.push(new Token(TokenType.MOD, ch, line, column));
      i++; column++; continue;
    }
    if (ch === '>') {
      tokens.push(new Token(TokenType.LEBIH_BESAR, ch, line, column));
      i++; column++; continue;
    }
    if (ch === '<') {
      tokens.push(new Token(TokenType.LEBIH_KECIL, ch, line, column));
      i++; column++; continue;
    }
    if (ch === '=') {
      tokens.push(new Token(TokenType.SAMA, ch, line, column));
      i++; column++; continue;
    }

    throw new Error(`Karakter tidak dikenal: '${ch}' di baris ${line}, kolom ${column}`);
  }

  tokens.push(new Token(TokenType.EOF, null, line, column));
  return tokens;
}

if (typeof module !== 'undefined') {
  module.exports = { TokenType, Token, tokenize };
}
